Warnings about deprecated `std::result_of` when building as part of Zeek

Here's another when building on macOS as part of Zeek:

```
In file included from /Users/tim/Desktop/projects/fmt/auxil/spicy/spicy/spicy/toolchain/src/compiler/codegen/production.cc:3:
In file included from /Users/tim/Desktop/projects/fmt/auxil/spicy/spicy/spicy/toolchain/include/spicy/compiler/detail/codegen/production.h:12:
In file included from /Users/tim/Desktop/projects/fmt/auxil/spicy/spicy/hilti/toolchain/include/hilti/global.h:8:
In file included from /Users/tim/Desktop/projects/fmt/auxil/spicy/spicy/hilti/toolchain/include/hilti/ast/ctor.h:7:
/Users/tim/Desktop/projects/fmt/auxil/spicy/spicy/hilti/toolchain/include/hilti/ast/node.h:893:29: warning: 'result_of<(lambda at /Users/tim/Desktop/projects/fmt/auxil/spicy/spicy/spicy/toolchain/src/compiler/codegen/production.cc:45:78) (hilti::expression::detail::Expression &)>' is deprecated [-Wdeprecated-declarations]
    using Y = typename std::result_of<F(X&)>::type;
                            ^
/Users/tim/Desktop/projects/fmt/auxil/spicy/spicy/spicy/toolchain/src/compiler/codegen/production.cc:44:68: note: in instantiation of function template specialization 'hilti::node::transform<hilti::expression::detail::Expression, (lambda at /Users/tim/Desktop/projects/fmt/auxil/spicy/spicy/spicy/toolchain/src/compiler/codegen/production.cc:45:78)>' requested here
                                    hilti::util::join(hilti::node::transform(x,
                                                                   ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/type_traits:4009:34: note: 'result_of<(lambda at /Users/tim/Desktop/projects/fmt/auxil/spicy/spicy/spicy/toolchain/src/compiler/codegen/production.cc:45:78) (hilti::expression::detail::Expression &)>' has been explicitly marked deprecated here
template <class _Callable> class _LIBCPP_DEPRECATED_IN_CXX17 result_of;
                                 ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__config:1066:39: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#  define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                      ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/__config:1043:48: note: expanded from macro '_LIBCPP_DEPRECATED'
#    define _LIBCPP_DEPRECATED __attribute__ ((deprecated))
                                               ^
1 warning generated.
```

`std::result_of` is deprecated in favor of `std::invoke_result` in C++17.

_Originally posted by @timwoj in https://github.com/zeek/spicy/issues/1188#issuecomment-1136608356_

--- a/hilti/runtime/include/util.h
+++ b/hilti/runtime/include/util.h
@@ -330,7 +330,7 @@ constexpr auto transform_result_value(const C&) {
 /** Applies a function to each element of container. */
 template<typename C, typename F>
 auto transform(const C& x, F f) {
-    using Y = typename std::result_of_t<F(typename C::value_type&)>;
+    using Y = typename std::invoke_result_t<F, typename C::value_type&>;
 
     auto y = detail::transform_result_value<C, Y>(x);
     std::transform(std::begin(x), std::end(x), std::inserter(y, std::end(y)), f);

--- a/hilti/toolchain/include/ast/node.h
+++ b/hilti/toolchain/include/ast/node.h
@@ -890,7 +890,7 @@ auto filter(const hilti::node::Set<X>& x, F f) {
  */
 template<typename X, typename F>
 auto transform(const hilti::node::Range<X>& x, F f) {
-    using Y = typename std::result_of<F(X&)>::type;
+    using Y = typename std::invoke_result_t<F, X&>;
     std::vector<Y> y;
     y.reserve(x.size());
     for ( const auto& i : x )
@@ -905,7 +905,7 @@ auto transform(const hilti::node::Range<X>& x, F f) {
  */
 template<typename X, typename F>
 auto transform(const hilti::node::Set<X>& x, F f) {
-    using Y = typename std::result_of<F(X&)>::type;
+    using Y = typename std::invoke_result_t<F, X&>;
     std::vector<Y> y;
     y.reserve(x.size());
     for ( const auto& i : x )

--- a/hilti/toolchain/include/base/util.h
+++ b/hilti/toolchain/include/base/util.h
@@ -86,7 +86,7 @@ using hilti::rt::transform; // NOLINT(misc-unused-using-decls)
 /** Applies a function to each element of a set, returning a vector with the results. */
 template<typename X, typename F>
 auto transform_to_vector(const std::set<X>& x, F f) {
-    using Y = typename std::result_of<F(X&)>::type;
+    using Y = typename std::invoke_result_t<F, X&>;
     std::vector<Y> y;
     y.reserve(x.size());
     for ( const auto& i : x )

